README.md
# Shared Systems for galactic war

Uses System Sharing and map packs to populate new Galactic Wars.

Select the system pool from Uber's systems, your systems, sharing servers, and installed map packs.

Use at your own risk - not all systems will be appropriate for the number of enemy commanders and subcommanders, and some are outright broken. That goes double for the default sharing server; the first one I tried the start planet collided instantly and the game couldn't start. If you find any such egregious systems, let me know and I'll add them to the blacklist.

Downloading from servers takes a bit. The default server has special case to sample 200 randomly from the first 2600 (the approximate total on 2016-03-17)


CHANGELOG.md
# Shared Systems for galactic war

## 1.2.5

- Rework system loading to better support map packs with asynchronous loading of system lists

## 1.2.4

- Add "Water World" to the list of broken systems

## 1.2.3

- Partial support for fromRandomList, as used in Galactic War AI Overhaul

## 1.2.2

- Handle explicit planets included in system templates, which may occur due to other mods, e.g. Galactic War AI Overhaul.

## 1.2.1

- Return a unique object from premade systems to prevent overwrite from the system being selected mutliple times.

## 1.2.0

- Remember selected system groups
- Deselect a server if system fetch fails, to prevent Go to War from getting left disabled
- Tooltips for remote icon and progress to explain multiplanet

## 1.1.0

- Add a loading indicator to system groups to better indicate why Go To War is disabled
- Add an icon to remote systems that may take a bit to download
- Added count of multiplanet systems in parenthesis

## 1.0.3

- Update title and tags

## 1.0.2

- Add a system to the naughty list

## 1.0.1

- If a mappack load fails, retry with lowercased filename
- Other improved error handling


ui/mods/gw_shared_systems/gw_start.js
(function() {
  model.systemSources = ko.observableArray([])

  model.persistedNames = ko.observable(["Uber"]).extend({local: 'gw_shared_systems_persisted_names'})

  model.selectedNames = ko.computed(function() {
    var names = []
    model.systemSources().forEach(function(opt) {
      if (opt.selected()) {
        names.push(opt.name)
      }
    })
    return names
  })

  $('#new-game-right .col-padding').append(
    '<div class="form-group">'+
      '<label id="system-sources-label" for="system-sources"><loc>Systems</loc></label>'+
      '<ul class="form-control" data-bind="foreach: systemSources">'+
        '<li class="system-group" data-bind="css: {selected: selected}">'+
          '<input type="checkbox" data-bind="attr: {id: id}, checked: selected"/>'+
          '<label data-bind="attr: {for: id}">'+
            '<span data-bind="text:name">Name</span> '+
            '<span data-bind="text:progress, tooltip: \'Downloaded/Total (Number of Multiplanet)\'"></span>'+
            '<img class="system-group-loading-indicator working" data-bind="visible:loading" src="coui://ui/main/shared/img/working.svg">'+
            '<img class="system-group-remote" data-bind="visible:remote, tooltip: \'Remote, may take some time to download\'" src="coui://ui/mods/gw_shared_systems/cloud-download.svg">'+
          '</label>'+
        '</li>'+
      '</ul>'+
    '</div>')

  requireGW(['main/game/galactic_war/shared/js/systems/template-loader'], function(chooseStarSystemTemplates) {
    chooseStarSystemTemplates.loadOptions().then(function(options) {
      model.selectedNames.subscribe(function(names) {
        chooseStarSystemTemplates.useSources(names)
        // force game build
        model.newGameSeed(Math.floor(Math.random() * 1000000).toString());
        model.persistedNames(names)
      })

      options.forEach(function(opt, i) {
        opt.selected = ko.observable(model.persistedNames().indexOf(opt.name) != -1)
        opt.id = 'system-group-'+i
      })
      model.systemSources(options)
    })
  })
})()


ui/mods/gw_shared_systems/map_packs.js
// !LOCNS:galactic_war
define([
], function () {

  var countMultiplanet = function(systems) {
    var multi = 0
    systems.forEach(function(system) {
      if (system.planets.length > 1) {
        multi++
      }
    })
    return " (" + multi + ")"
  }

  var fixupPlanetConfig = function (system) {
    UberUtility.fixupPlanetConfig(system)

    system.surface_area = 0
    system.planets.forEach(function(planet) {
      if (planet.generator && planet.generator.biome != 'gas') {
        system.surface_area += 4 * Math.PI * Math.pow(planet.generator.radius, 2) * 0.000001
      }
    })
    return system;
  }

  var mapPacks = {}

  window.cShareSystems = window.cShareSystems || {}
  cShareSystems.load_pas = function(tabName, fileArray) {
    if (mapPacks[tabName]) {
      mapPacks[tabName].files = fileArray
    } else {
      mapPacks[tabName] = {
        files: fileArray,
        requested: [],
        systems: [],
        promise: null
      }
    }
  }

  var loadPack = function(tabName, progress) {
    if (mapPacks[tabName].promise) {
      return mapPacks[tabName].promise
    }
    mapPacks[tabName].promise = $.Deferred()
    var loadFiles = function() {
      var fileArray = mapPacks[tabName].files
      var fileCountSnapshot = fileArray.length
      if (fileCountSnapshot < 1) return setTimeout(loadFiles, 1000)
      var requested = mapPacks[tabName].requested
      var systems = mapPacks[tabName].systems
      var counter = 0

      var done = function() {
        counter--;

        progress(systems.length+'/'+fileCountSnapshot + countMultiplanet(systems))

        if (counter <= 0) {
          if (mapPacks[tabName].files.length > fileCountSnapshot) return setTimeout(loadFiles, 1000)
          systems.forEach(fixupPlanetConfig)
          mapPacks[tabName].promise.resolve(systems)
        }
      }

      progress('0/'+fileCountSnapshot)

      for (arrayItem in fileArray) {
        if (requested[arrayItem]) continue;
        requested[arrayItem] = true;
        counter++
        var fileName = fileArray[arrayItem];
        $.getJSON(fileName).then(function(data) {
          systems.push(data);
          done()
        }, function(xhr, err, ex) {
          if (xhr.status == 404) {
            $.getJSON(fileName.toLowerCase()).then(function(data) {
              systems.push(data);
            }).always(done)
          } else {
            done()
          }
        })
      }
    }

    loadFiles()

    return mapPacks[tabName].promise
  };

  var mapPacksLoaded

  var mapPackList = function() {
    if (mapPacksLoaded) {
      return mapPacksLoaded
    }
    mapPacksLoaded = $.Deferred()
    $.when.apply($, (scene_mod_list.load_planet || []).map(function(path) {
      var promise = $.Deferred()
      $.get(path, null, null, 'text').then(function(contents) {
        if (contents.match(/cShareSystems.load_pas\s*\(/)) {
          var se = document.createElement('script');
          se.type = "text/javascript";
          se.text = contents;
          document.getElementsByTagName('head')[0].appendChild(se);
        }
      }).always(function() {promise.resolve()})
      return promise
    })).always(function() {
      mapPacksLoaded.resolve(mapPacks)
    })
    return mapPacksLoaded
  }

  return {
    mapPackList: mapPackList,
    loadPack: loadPack
  }
});


ui/mods/gw_shared_systems/user_systems.js
// !LOCNS:galactic_war
define([
], function () {

  var countMultiplanet = function(systems) {
    var multi = 0
    systems.forEach(function(system) {
      if (system.planets.length > 1) {
        multi++
      }
    })
    return " (" + multi + ")"
  }

  var fixupPlanetConfig = function (system) {
    UberUtility.fixupPlanetConfig(system)

    system.surface_area = 0
    system.planets.forEach(function(planet) {
      if (planet.generator && planet.generator.biome != 'gas') {
        system.surface_area += 4 * Math.PI * Math.pow(planet.generator.radius, 2) * 0.000001
      }
    })
    return system;
  }

  var userSystems = ko.observableArray([]).extend({ db: { local_name: 'systems', db_name: 'misc' }});

  userSystems.subscribe(function(systems) {
    systems.forEach(fixupPlanetConfig)
  })

  var waitForPlanetToLoad = function (planet_spec) {
    var deferred = $.Deferred();

    UberUtility.waitForAttributeLoad(planet_spec, 'csg_key', 'planetCSG', constants.PLANET_CSG_DATABASE).then(function (first) {
      UberUtility.waitForAttributeLoad(first, 'metal_spots_key', 'metal_spots', constants.PLANET_METAL_SPOTS_DATABASE).then(function (second) {
        UberUtility.waitForAttributeLoad(second, 'landing_zones_key', 'landing_zones', constants.PLANET_LANDING_ZONES_DATABASE).then(function (third) {
          _.omit(third, 'source');
          deferred.resolve(third);
        });;
      });
    });

    return deferred.promise();
  };

  var waitForSystemToLoad = function (system, options /* { omit_keys } */) {
    var deferred = $.Deferred();
    var array = _.map(system.planets, waitForPlanetToLoad);

    UberUtility.waitForAll(array).then(function (results) {
      system.planets = results;

      if (options.omit_keys)
        system.planets = _.map(system.planets, function (element) {
          return _.omit(element, ['csg_key', 'metal_spots_key', 'landing_zones_key']);
        });

        deferred.resolve(system);
    });

    return deferred.promise();
  };

  var systemsLoaded

  var load = function(progress) {
    if (systemsLoaded) {
      return systemsLoaded
    }
    progress('0/'+userSystems().length)
    systemsLoaded = $.Deferred()
    $.when(userSystems().map(waitForSystemToLoad)).then(function() {
      progress(userSystems().length+'/'+userSystems().length + countMultiplanet(userSystems()))
      systemsLoaded.resolve(userSystems())
    })
    return systemsLoaded
  }

  return {
    systems: userSystems,
    load: load,
  }
});


ui/mods/gw_shared_systems/template_builder.js
// !LOCNS:galactic_war
define([
], function () {
  var planet_template = {
    name: "Default Planet",
    mass: 5000,
    position: [0, 0],
    velocity: [0, 0],
    required_thrust_to_move: 0,
    generator: {
      seed: 15,
      radius: 100,
      heightRange: 25,
      waterHeight: 35,
      temperature: 100,
      metalDensity: 50,
      metalClusters: 50,
      biomeScale: 100,
      biome: "earth"
    }
  };

  return function(config) {
    //console.log('generate template', config)
    var rng = new Math.seedrandom(config.seed !== undefined ? config.seed : Math.random());

    var getRandomInt = function (min, max) {
      return Math.floor(rng() * (max - min + 1)) + min;
    };

    var rSystem = {
      name: config.name || ("PA-" + getRandomInt(100, 30000)),
      description: '',
      isRandomlyGenerated: true
    };

    var cSys = _.cloneDeep(config.template);
    rSystem.planets = _.map(cSys.Planets, function(plnt, index) {
      //console.log(plnt);
      if (plnt.fromRandomList && plnt.fromRandomList.planets) {
        var choice = _.cloneDeep(plnt.fromRandomList.planets[getRandomInt(0, plnt.fromRandomList.planets.length-1)])
        var overrides = _.cloneDeep(plnt)
        delete overrides.fromRandomList
        _.assign(choice, overrides)
        plnt = choice
      }
      if (plnt.isExplicit) {
        return _.cloneDeep(plnt);
      }
      var bp = _.cloneDeep(planet_template);
      bp.generator.seed = getRandomInt(0, 32767);
      bp.generator.biome = _.sample(plnt.Biomes);

      bp.generator.radius = getRandomInt(plnt.Radius[0], plnt.Radius[1])
      bp.generator.heightRange = getRandomInt(plnt.Height[0], plnt.Height[1]);
      bp.generator.waterHeight = getRandomInt(plnt.Water[0], plnt.Water[1]);
      bp.generator.waterDepth = 100;
      bp.generator.temperature = getRandomInt(plnt.Temp[0], plnt.Temp[1]);
      bp.generator.biomeScale = getRandomInt(plnt.BiomeScale[0], plnt.BiomeScale[1]);
      bp.generator.metalDensity = getRandomInt(plnt.MetalDensity[0], plnt.MetalDensity[1]);
      bp.generator.metalClusters = getRandomInt(plnt.MetalClusters[0], plnt.MetalClusters[1]);
      bp.generator.index = index;
      bp.name = plnt.name;
      bp.position = plnt.Position;
      bp.velocity = plnt.Velocity;
      bp.required_thrust_to_move = getRandomInt(plnt.Thrust[0], plnt.Thrust[1]);
      bp.mass = plnt.mass;
      bp.starting_planet = plnt.starting_planet;

      return bp;
    });

    return rSystem
  };
});


ui/main/game/galactic_war/shared/js/systems/template-loader.js
// !LOCNS:galactic_war
define([
  '/mods/gw_shared_systems/shared_systems.js',
  '/mods/gw_shared_systems/map_packs.js',
  '/mods/gw_shared_systems/template_builder.js',
  '/main/shared/js/premade_systems.js',
  '/mods/gw_shared_systems/user_systems.js',
], function (sharedSystems, mapPacks, generateFromTemplate, premade, user) {
  /*
  requireGW(['main/game/galactic_war/shared/js/systems/titans-normal'], function(temp) {
    temp.forEach(function(set) {
      console.log(set.Players)
      set.Systems.forEach(function(system) {
        min = 0
        max = 0
        system.Planets.forEach(function(planet) {
          if (planet.Biomes[0] != 'gas') {
            min += 4 * Math.PI * Math.pow(planet.Radius[0], 2)
            max += 4 * Math.PI * Math.pow(planet.Radius[1], 2)
          }
        })
        console.log(min/1000000, max/1000000)
      })
    })
  })
  */

  var biomes = [
    'earth', 'desert', 'lava', 'metal', 'moon', 'tropical', 'gas',
    '1v1test', 'asteroid', 'csg_debug', 'ice_boss', 'metal_boss', 'sandbox',
  ]

  var nameBlocklist = [
    'flint and steel', // start planets immediately collide
    'Molten Metal', // bugged pathing, no metal
    'Water World', // reported to have no spawn locations
  ]

  var complexBlocklist = [
    _.matches({ // all planets immediately collide
      creator: 'octobomb',
      planets: [
        {name: 'Footfall'},
        {name: 'Mariani'},
      ]
    }),
  ]

  var countMultiplanet = function(systems) {
    var multi = 0
    systems.forEach(function(system) {
      if (system.planets.length > 1) {
        multi++
      }
    })
    return " (" + multi + ")"
  }

  var withoutBrokenSystems = function(systems) {
    return systems.filter(function(system) {
      if (nameBlocklist.indexOf(system.name) != -1) return false
      for (var b in complexBlocklist) {
        if (complexBlocklist[b](system)) return false
      }

      var startingPlanets = 0
      for (var i in system.planets) {
        var planet = system.planets[i]
        if (planet.starting_planet) {
          startingPlanets++
        }
        if (biomes.indexOf(planet.generator.biome) == -1) {
          return false
        }
      }

      if (startingPlanets < 1) {
        system.planets[0].starting_planet = true
      }

      return true
    })
  }

  var fixupPlanetConfig = function (system) {
    UberUtility.fixupPlanetConfig(system)

    system.surface_area = 0
    system.planets.forEach(function(planet) {
      if (planet.generator && planet.generator.biome != 'gas') {
        system.surface_area += 4 * Math.PI * Math.pow(planet.generator.radius, 2) * 0.000001
      }
    })
    return system;
  }

  premade.forEach(fixupPlanetConfig)

  var loadSelectedSources = function(choices) {
    var systemsLoaded = $.Deferred()

    loadOptions().then(function(options) {
      var loading = []
      choices.forEach(function(name) {
        var it = _.find(options, 'name', name)
        if (it) {
          it.loading(true)
          loading.push(it.load()
                        .always(function() {it.loading(false)})
                        .fail(function() { if (it.selected) {it.selected(false)}})
                      )
        }
      })
      $.when.apply($, loading).then(function() {
        if (arguments.length > 0) {
          systemsLoaded.resolve(_.flatten(arguments))
        } else {
          systemsLoaded.reject()
        }
      })
    })

    return systemsLoaded
  }

  var chooseStarSystemTemplates = function(content, easier) {
    //console.log('create')
    var systemsLoaded = loadSelectedSources(choices)

    /*
    systemsLoaded.then(function(sys) {
      console.log(sys.sort(function(a, b) {
        return a.surface_area - b.surface_area
      }).map(function(sys) {return Math.floor(sys.surface_area)}))
    })
    */

    var generate = function(config) {
      //console.log('generate', config)

      var pickSystem = function(systems) {
        systems = withoutBrokenSystems(systems)
        //console.log(systems)
        var rng = new Math.seedrandom(config.seed !== undefined ? config.seed : Math.random());

        var getRandomInt = function (min, max) {
          return Math.floor(rng() * (max - min + 1)) + min;
        };

        if (systems.length > 0) {
          var min = config.players*0.5
          var max = config.players*4
          var candidates = systems.filter(function(s) {
            return min < s.surface_area && s.surface_area < max
          })
          if (candidates.length < 1) {
            candidates = systems
          }
          var i = getRandomInt(0, candidates.length - 1)
          //console.log(i, candidates.length)
          //console.log('generate', candidates[i].surface_area)
          return JSON.parse(JSON.stringify(candidates[i]))
        }
      }

      if (config.template) {
        var p = $.Deferred()
        p.resolve(generateFromTemplate(config))
        return p
      } else {
        return systemsLoaded.then(pickSystem)
      }
    };

    return {
      generate: generate
    };
  };

  var userProgress = ko.observable('')
  var baseOptions = [
    {
      name: 'Uber',
      remote: false,
      progress: ko.observable(premade.length.toString() + countMultiplanet(premade)),
      loading: ko.observable(false),
      load: function() {
        var promise = $.Deferred()
        promise.resolve(premade)
        return promise
      },
    },
    {
      name: 'My Systems',
      remote: false,
      progress: userProgress,
      loading: ko.observable(false),
      load: function() {
        return user.load(userProgress)
      },
    },
  ]

  var optionsPromise

  var loadOptions = function() {
    if (optionsPromise) return optionsPromise

    var options = _.cloneDeep(baseOptions)
    //var serverPromise = $.Deferred(); serverPromise.resolve(true)
    var serverPromise = sharedSystems.getServerList().then(function(servers) {
      servers.forEach(function(server) {
        var progress = ko.observable('')
        options.push({
          name: server.name,
          remote: true,
          progress: progress,
          loading: ko.observable(false),
          load: function() {
            return sharedSystems.loadSystems(server.search_url, 200, progress)
          },
        })
      })
    })

    var packPromise = mapPacks.mapPackList().then(function(packs) {
      Object.keys(packs).forEach(function(name) {
        var progress = ko.observable('')
        options.push({
          name: name,
          remote: false,
          progress: progress,
          loading: ko.observable(false),
          load: function() {
            return mapPacks.loadPack(name, progress)
          },
        })
      })
    })

    optionsPromise = $.when(serverPromise, packPromise).then(function() {
      return options
    })
    return optionsPromise
  }

  var choices = []
  var useSources = function(names) {
    choices = names
  }

  chooseStarSystemTemplates.loadOptions = loadOptions
  chooseStarSystemTemplates.useSources = useSources

  return chooseStarSystemTemplates;
});


ui/mods/gw_shared_systems/shared_systems.js
// !LOCNS:galactic_war
define([
], function () {

  var countMultiplanet = function(systems) {
    var multi = 0
    systems.forEach(function(system) {
      if (system.planets.length > 1) {
        multi++
      }
    })
    return " (" + multi + ")"
  }

  var fixupPlanetConfig = function (system) {
    UberUtility.fixupPlanetConfig(system)

    system.surface_area = 0
    system.planets.forEach(function(planet) {
      if (planet.generator && planet.generator.biome != 'gas') {
        system.surface_area += 4 * Math.PI * Math.pow(planet.generator.radius, 2) * 0.000001
      }
    })
    return system;
  }

  var serverListUrl = "https://raw.githubusercontent.com/pamods/mods-conundrum/master/cShareSystems_serverList/serverlist.json";

  var defaultServer = "http://1-dot-winged-will-482.appspot.com/search"

  var serverOptions = [
    {
      "name"		: "Default Server",
      "save_url"	: "http://1-dot-winged-will-482.appspot.com/save",
      "search_url"	: "http://1-dot-winged-will-482.appspot.com/search"
      //"search_url"	: "http://1-dot-realm-sharing.appspot.com/search"
    }
  ]

  var getServerList = function() {
    return $.getJSON(serverListUrl).then(function(data) {
      if (!data || !data.servers) return;

      serverOptions = data.servers
      return serverOptions
    });
  };

  var filterOptions = {
    "name"			: "",
    "creator"		: "",
    "minPlanets"	: 1,
    "maxPlanets"	: 16,
    "sort_field"	: "system_id",
    "sort_direction": "DESC",
    "limit"			: 16
  }

  var searchSystems = function(search_url, parameters) {
    var request = $.Deferred()
    parameters.request_time = Date.now();
    $.get(search_url, parameters, null, 'json')
      .then(function(data) {
        if (typeof data == "string") {
          var message = (typeof data == "string");
          request.reject(message)
        } else {
          request.resolve(data)
        }
      }, function(jqXHR, textStatus, errorThrown) {
        request.reject(textStatus, errorThrown)
      })
    return request
  }

  var systemsLoaded = {}

  var loadSystems = function(search_url, fetchLimit, progress) {
    search_url = search_url || serverOptions[0].search_url
    if (systemsLoaded[search_url]) {
      return systemsLoaded[search_url]
    }

    console.log('load')
    var promise = systemsLoaded[search_url] = $.Deferred()
    var systems = []
    progress(systems.length+'/'+fetchLimit + countMultiplanet(systems))
    var next = function(start) {
      if (search_url == defaultServer) {
        //approximate number 2016-03-17
        start = Math.floor(Math.random()*2600)
      }
      console.log('search', start)
      searchSystems(search_url, $.extend({start: start}, filterOptions))
        .then(function(data) {
          //console.log(data)
          systems = systems.concat(data.systems)
          progress(systems.length+'/'+fetchLimit + countMultiplanet(systems))
          if (systems.length >= fetchLimit) {
            console.warn(systems.length, 'thats crazy')
            progress(systems.length+'/'+systems.length + countMultiplanet(systems))
            promise.resolve(systems.map(fixupPlanetConfig))
            return
          }
          if (data.total > start + filterOptions.limit) {
            next(start + filterOptions.limit)
          } else {
            console.log('no more', data.total)
            progress(systems.length+'/'+systems.length + countMultiplanet(systems))
            promise.resolve(systems.map(fixupPlanetConfig))
            return
          }
        }, function() {
          console.error('fetch failed', start)
          if (systems.length > 0) {
            progress(systems.length+'/'+systems.length + countMultiplanet(systems))
            promise.resolve(systems.map(fixupPlanetConfig))
          } else {
            progress('0')
            promise.reject(arguments)
          }
        })
    }
    next(0)
    return promise
  }

  return {
    getServerList: getServerList,
    loadSystems: loadSystems,
  }
});
